### ISSUE TITLE
WCAG issue (3.3.2: Labels or Instructions)

### ISSUE THREAD

**Issue by olavflar** [digdir/baksia#923]
The issue is related to the search field on the top of the page in the header of all the pages 

Placeholder used for label 
Description: The form control only uses placeholder="..." to provide a visible label.

<input class="pagefind-ui__search-input svelte-e9gkc3" type="text" placeholder="Søk" autocapitalize="none" enterkeyhint="search">

**Comment by altinnadmin** [comment]
> The form control only uses placeholder="..." to provide a visible label.

@sorensensig @olavflar I don't agree, so we should discuss this :) 

The search is rocking quite a few accessibility-features that the tool you're using seems to have missed:

1. The form element for search has attribute [role="search"](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/search_role)
   - So the semantics and easy navigation to search is communicated to screen readers
2. The form element for search also has [aria-label="Søk på denne siden"](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Attributes/aria-label)
   - So a label *is* provided for screen readers
3. The search input uses [enterkeyhint="search"](https://developer.mozilla.org/en-US/docs/Web/HTML/Global_attributes/enterkeyhint)
   - Giving the user agent information that can be used to make it clearer for the user that a search will be performed, f.ex. by augmenting virtual keyboards or giving hints in screen readers
4. The search input field has [placeholder="Søk"](https://developer.mozilla.org/en-US/docs/Web/HTML/Attributes/placeholder)
   - Giving a visual hint to the user that this is a search
5. The search input field also has magnifier icon 🔍 to the left
   - Giving another visual hint to the user that this is a search
6. A clean and minimalistic design has a value by itself

**Comment by sorensensig** [comment]
@altinnadmin, the accessibility features you are listing are great and helps most people with accessibility needs. The github issue is somewhat vague, so I'll offer a few examples to better explain the part that is still missing :).

The WCAG issue stems from using the placeholder text in place of a label. Placeholder texts have a tendency to disappear when a user inputs text, meaning that the context or action (affordance) disappears as you start to type. Conversely, a label does not disappear when you start to type text, meaning that the context/affordance will stick around. This can be particularly important for people struggling with short term memory or attention span.

Example of how AirBnB use both a label and placeholder text to inform the user about what the context/affordance of the search field. Pay attention to the label "where", telling the user to input a destination.
<img width="874" alt="Screenshot 2024-10-30 at 14 27 46" src="https://github.com/user-attachments/assets/530232eb-04cd-4f99-a2f2-e2000f2e4873">
<img width="877" alt="Screenshot 2024-10-30 at 14 34 38" src="https://github.com/user-attachments/assets/794adf6a-e16b-42f0-8330-71c91af10a31">

Example of how the users looses the context when typing in the search field on Baksia. Pay attention to how the word "søk", telling the user what to do disappears on input.
<img width="877" alt="Screenshot 2024-10-30 at 14 34 24" src="https://github.com/user-attachments/assets/4bfc5712-46b5-4108-9ca9-67a849c7bdad">
<img width="877" alt="Screenshot 2024-10-30 at 14 34 28" src="https://github.com/user-attachments/assets/4a1b042b-cf93-421c-85df-7dcff893df35">

Hope this makes sense :).

**Comment by altinnadmin** [comment]
> Example of how AirBnB use both a label and placeholder 

Yes, this makes sense for actual forms, where a user might loose context and get lost over time.

> Example of how the users looses the context when typing in the search field on Baksia. Pay attention to how the word "søk", telling the user what to do disappears on input.

But it makes no sense to just show the input field of Baksia in isolation. Then you're artificially hiding the context.

In Baksia, the very moment the user enters the first character, the result is there instantly, with reference to the character you just typed. And the search icon is still there.

![image](https://github.com/user-attachments/assets/5d5fad10-924f-44d6-a65f-6e62bb1ba843)

![image](https://github.com/user-attachments/assets/fac771e1-3582-46ca-b5af-2c53491c485f)

A "somewhat" known solution for search does something very similar... :) 

![image](https://github.com/user-attachments/assets/cdcde628-049f-4a82-b13c-8df47040301f)


And to make it even more obvious how strongly the visual context of "search in Baksia" is communicated to the user (this time on mobile):

![image](https://github.com/user-attachments/assets/5655f907-06aa-4ddb-b73e-014b7c11883d)

**Comment by sorensensig** [comment]
From a user experience perspective I agree with your point of view. It does feel somewhat out of place to force a label into a search field that offers immediate results.

I've looked at the standard, both the w3 version and the one from uutilsynet. The w3 standard is quite clear about the criterion covering all input elements: "Labels or instructions are provided when content requires user input." ([ref](https://www.w3.org/WAI/WCAG22/Understanding/labels-or-instructions.html#:~:text=Labels%20or%20instructions%20are%20provided%20when%20content%20requires%20user%20input.)). However, the uutilsynet one refers to form elements rather than all input elements ([ref](https://www.uutilsynet.no/wcag-standarden/332-ledetekster-eller-instruksjoner-niva/117#:~:text=KORT%20FORTALT%3A%20Det%20vises%20ledetekster%20eller%20instruksjoner%20n%C3%A5r%20du%20har%20skjemaelementer%20som%20m%C3%A5%20fylles%20ut.)).

However, none of the above specifically mention search components. Looking at these two sources ([ref](https://www.w3.org/WAI/WCAG22/quickref/?showtechniques=332#labels-or-instructions:~:text=F82%3A%20Failure%20of%20Success%20Criterion%203.3.2%20by%20visually%20formatting%20a%20set%20of%20phone%20number%20fields%20but%20not%20including%20a%20text%20label))([ref](https://design.education.gov.uk/accessibility/understanding-wcag/understandable#:~:text=When%20data%20must%20be%20entered%20in%20a%20specific%20format%20or%20in%20a%20particular%20way%2C%20clear%20instructions%20must%20be%20associated%20with%20the%20form%20field.)), they seem to indicate that the criterion is intended for form fields with specific input formats. Although the gov uk one  still points to the same solution "All form fields that expect user input/interaction must have a label." ([ref](https://design.education.gov.uk/accessibility/understanding-wcag/understandable#:~:text=All%20form%20fields%20that%20expect%20user%20input/interaction%20must%20have%20a%20label.)).

**Other sources**
[3.3.2 labels and instructions](https://www.wcag.com/developers/3-3-2-labels-instructions/)
[Understanding SC 3.3.2](https://www.w3.org/WAI/WCAG21/Understanding/labels-or-instructions.html)

It's worth mentioning that this is a A-level requirement, meaning that it's the strictest and most basic of the three types. However, I can't seem to find content online about how to evaluate search components specifically. I also can't seem to find any critique of google missing a label on their search field. As for the example of Airbnb, it makes sense in their instance to have a label as they are asking for a specific type of input (destination).

**Conclusion**
I think we ought to follow the guideline of gov.uk on this one, meaning: "When data must be entered in **a specific format or in a particular way**, clear instructions must be associated with the form field.". Let's follow your gut instinct and avoid using a label for the global search in this instance. Closing the issue :).

**Comment by altinnadmin** [comment]
@sorensensig Since I had to fix contrast anyways, I added aria-label to the input. That's what google was doing. So now we're double-good :-P

![image](https://github.com/user-attachments/assets/8ebf30e5-bd93-45ad-b826-461a1732ab42)
